src/invert/laplacexy/laplacexy.cxx

```diff
@@ -105,96 +164,251 @@ LaplaceXY::LaplaceXY(Mesh *m, Options *opt, const CELL_LOC loc)
   PetscMalloc( (localN)*sizeof(PetscInt), &d_nnz );
   PetscMalloc( (localN)*sizeof(PetscInt), &o_nnz );
   
-  for(int i=0;i<localN;i++) {
-    // Non-zero elements on this processor
-    d_nnz[i] = 5; // Star pattern in 2D
-    // Non-zero elements on neighboring processor
-    o_nnz[i] = 0;
-  }
-  
-  // X boundaries
-  if(localmesh->firstX()) {
-    // Lower X boundary
-    for(int y=localmesh->ystart;y<=localmesh->yend;y++) {
-      const int localIndex = globalIndex(localmesh->xstart - 1, y);
-      ASSERT1( (localIndex >= 0) && (localIndex < localN) );
-    
-      d_nnz[localIndex] = 2; // Diagonal sub-matrix
-      o_nnz[localIndex] = 0; // Off-diagonal sub-matrix
+  if (finite_volume) {
```

Comment:
This whole bit is now a bit unwieldy. Worth pulling out into two helper functions? Then could just be

```cpp
if (finite_volume) {
   set_matrix_elements_for_finite_volume();
} else {
   set_matrix_elements_for_finite_difference();
}
```